$OpenBSD: patch-libpurple_protocols_yahoo_libymsg_c,v 1.1.2.1 2011/03/13 17:37:46 jasper Exp $

Security fix for CVE-2011-1091, improper handling of malformed YMSG packets
cause NULL pointer dereferences, leading to denial of service.

Patch from upstream git:
http://developer.pidgin.im/viewmtn/revision/info/a7c415abba1f5f01f79295337518837f73d99bb7

--- libpurple/protocols/yahoo/libymsg.c.orig	Sun Oct 17 05:58:01 2010
+++ libpurple/protocols/yahoo/libymsg.c	Sun Mar 13 18:29:10 2011
@@ -842,7 +842,7 @@ static void yahoo_process_notify(PurpleConnection *gc,
 				break;
 		}
 
-		if (*stat == '1')
+		if (stat && *stat == '1')
 			serv_got_typing(gc, fed_from, 0, PURPLE_TYPING);
 		else
 			serv_got_typing_stopped(gc, fed_from);
@@ -864,7 +864,7 @@ static void yahoo_process_notify(PurpleConnection *gc,
 
 		yahoo_friend_set_game(f, NULL);
 
-		if (*stat == '1') {
+		if (stat && *stat == '1') {
 			yahoo_friend_set_game(f, game);
 			if (bud)
 				yahoo_update_status(gc, from, f);
@@ -920,6 +920,11 @@ static void yahoo_process_sms_message(PurpleConnection
 		if (pair->key == 16)
 			server_msg = pair->value;
 		l = l->next;
+	}
+
+	if(!sms) {
+		purple_debug_info("yahoo", "Received a malformed SMS packet!\n");
+		return;
 	}
 
 	if( (pkt->status == -1) || (pkt->status == YAHOO_STATUS_DISCONNECTED) ) {
